File: How to write an SFTP file transfer client for WinForms/[VB]-How to write an SFTP file transfer client for WinForms/VB/ClientController_List.vb
Imports System.IO
Imports System.Text
Imports ComponentPro
Imports ComponentPro.IO
Imports ComponentPro.Net

Namespace ClientSample
	Partial Friend Class ClientController
		#Region "Remote"

		''' <summary>
		''' Refreshes the remote list view.
		''' </summary>
#If Framework4_5 Then
		Public Async Sub RefreshRemoteList()
#Else
		Public Sub RefreshRemoteList()
#End If
			' Disable the dialog.
			_view.EnableView(False)

#If Framework4_5 Then
			Try
				' Asynchronously retrieve a list of remote files.
#If Framework4_5 Then
				Dim files As FtpFileInfoCollection = Await client.ListDirectoryAsync()
#Else
				Dim files As FtpFileInfoCollection = client.ListDirectory()
#End If
				ShowRemoteList(files)
			Catch exc As Exception
				If Not HandleException(exc) Then
					Return
				End If
			Finally
				' Enable the dialog and handle the exception.
				EnableDialog(True)
			End Try
#Else
			_client.ListDirectoryAsync()
#End If
		End Sub

#If Not Framework4_5 Then
		''' <summary>
		''' Handles the client's GetFileListCompleted event.
		''' </summary>
		''' <param name="sender">The Ftp object.</param>
		''' <param name="e">The event arguments.</param>
		Private Sub client_ListDirectoryCompleted(ByVal sender As Object, ByVal e As ExtendedAsyncCompletedEventArgs(Of FileInfoCollection))
			_view.EnableView(True)

			If e.Error IsNot Nothing Then
				If Not HandleException(e.Error) Then
					Return
				End If
				Disconnect()
				Return
			End If

			DoListDirPost(e.Result)
		End Sub
#End If

		Private Sub DoListDirPost(ByVal files As FileInfoCollection)
			ListFiles(files, _currentDirectory.Length <= 1, False)
		End Sub

		Private Sub ListFiles(ByVal files As FileInfoCollection, ByVal rootDir As Boolean, ByVal local As Boolean)
			Dim viewList As New List(Of ListItemInfo)()
			Dim item As ListItemInfo

			' Add directories into the list first.
			For i As Integer = 0 To files.Count - 1
				Dim f As AbstractFileInfo = files(i)

				If f.IsDirectory Then
					If f.Name <> "." AndAlso f.Name <> ".." Then
						item = New ListItemInfo()
						item.FileInfo = f

						If Not local Then
							item.Permissions = _clientPlugin.GetPermissionsString(f)
						End If

						viewList.Add(item)
					End If
				ElseIf f.IsSymlink Then ' Add symlinks.
					item = New ListItemInfo()
					item.FileInfo = f
					item.LinkedFile = _client.GetFileInfo(f.SymlinkPath)
					item.Permissions = _clientPlugin.GetPermissionsString(f)

					viewList.Add(item)
				Else 'Add Files.
					item = New ListItemInfo()
					item.FileInfo = f
					If Not local Then
						item.Permissions = _clientPlugin.GetPermissionsString(f)
					End If

					viewList.Add(item)
				End If
			Next i

			If Not rootDir Then
				' Add Cdup list item.
				item = New ListItemInfo()
				If local Then
					item.FileInfo = DiskFileSystem.Default.CreateFileInfo("..", FileAttributes.Directory, -1, Date.MinValue)
				Else
					item.FileInfo = _client.CreateFileInfo("..", FileAttributes.Directory, -1, Date.MinValue)
				End If
				viewList.Add(item)
			End If

			If local Then
				_view.ListLocalFiles(viewList)
			Else
				_view.ListRemoteFiles(viewList)
			End If
		End Sub

		#End Region

		#Region "Local"

		''' <summary>
		''' Refreshes the local list view.
		''' </summary>
		Public Sub RefreshLocalList()
			Try
				Dim directory As New DirectoryInfo(_currentLocalDirectory)
				Dim list As FileInfoCollection = DiskFileSystem.Default.ListDirectory(directory.FullName)

				ListFiles(list, directory.FullName.Length <= 3, True)
			Catch exc As Exception
				_view.ShowError(exc)
				Return
			End Try
		End Sub

		#End Region

		''' <summary>
		''' Changes local directory.
		''' </summary>
		''' <param name="dir">The new target path.</param>
		Public Sub DoRemoteChangeDirectory(ByVal dir As String)
			If _state <> ConnectionState.Ready OrElse dir = "." Then
				Return
			End If

			If dir = ".." Then
				dir = _client.GetDirectoryName(_currentDirectory)
			Else
				dir = _client.CombinePath(_currentDirectory, dir)
			End If

			Try
				_client.SetCurrentDirectory(dir)
			Catch ex As Exception
				_view.ShowError(ex)
				Return
			End Try

			_view.UpdateRemotePath(dir)
			_currentDirectory = dir
			RefreshRemoteList()
		End Sub

		''' <summary>
		''' Changes local directory.
		''' </summary>
		''' <param name="dir">The new target path.</param>
		Public Sub DoLocalChangeDirectory(ByVal dir As String)
			dir = Path.GetFullPath(Path.Combine(_currentLocalDirectory, dir))
			_view.UpdateLocalPath(dir)
			_currentLocalDirectory = dir
			RefreshLocalList()
		End Sub
	End Class
End Namespace

File: How to write an SFTP file transfer client for WinForms/[C#]-How to write an SFTP file transfer client for WinForms/C#/ClientController_List.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Windows.Forms;
using ComponentPro;
using ComponentPro.IO;
using ComponentPro.Net;

namespace ClientSample
{
    partial class ClientController
    {
        #region Remote

        /// <summary>
        /// Refreshes the remote list view.
        /// </summary>
        public
#if Framework4_5
        async
#endif
        void RefreshRemoteList()
        {
            // Disable the dialog.
            _view.EnableView(false);

#if Framework4_5
            try
            {
                // Asynchronously retrieve a list of remote files.
#if Framework4_5
                FtpFileInfoCollection files = await client.ListDirectoryAsync();
#else
                FtpFileInfoCollection files = client.ListDirectory();
#endif
                ShowRemoteList(files);
            }
            catch (Exception exc)
            {
                if (!HandleException(exc))
                    return;
            }
            finally
            {
                // Enable the dialog and handle the exception.
                EnableDialog(true);
            }
#else
            _client.ListDirectoryAsync();
#endif
        }

#if !Framework4_5
        /// <summary>
        /// Handles the client's GetFileListCompleted event.
        /// </summary>
        /// <param name="sender">The Ftp object.</param>
        /// <param name="e">The event arguments.</param>
        private void client_ListDirectoryCompleted(object sender, ExtendedAsyncCompletedEventArgs<FileInfoCollection> e)
        {
            _view.EnableView(true);

            if (e.Error != null)
            {
                if (!HandleException(e.Error))
                    return;
                Disconnect();
                return;
            }

            DoListDirPost(e.Result);
        }
#endif

        void DoListDirPost(FileInfoCollection files)
        {            
            ListFiles(files, _currentDirectory.Length <= 1, false);
        }

        private void ListFiles(FileInfoCollection files, bool rootDir, bool local)
        {
            List<ListItemInfo> viewList = new List<ListItemInfo>();
            ListItemInfo item;

            // Add directories into the list first.
            for (int i = 0; i < files.Count; i++)
            {
                AbstractFileInfo f = files[i];

                if (f.IsDirectory)
                {
                    if (f.Name != "." && f.Name != "..")
                    {
                        item = new ListItemInfo();
                        item.FileInfo = f;

                        if (!local)
                            item.Permissions = _clientPlugin.GetPermissionsString(f);

                        viewList.Add(item);
                    }
                }
                else if (f.IsSymlink) // Add symlinks.
                {
                    item = new ListItemInfo();
                    item.FileInfo = f;
                    item.LinkedFile = _client.GetFileInfo(f.SymlinkPath);
                    item.Permissions = _clientPlugin.GetPermissionsString(f);

                    viewList.Add(item);
                }
                else //Add Files.
                {
                    item = new ListItemInfo();
                    item.FileInfo = f;
                    if (!local)
                        item.Permissions = _clientPlugin.GetPermissionsString(f);

                    viewList.Add(item);
                }
            }

            if (!rootDir)
            {
                // Add Cdup list item.
                item = new ListItemInfo();
                if (local)
                    item.FileInfo = DiskFileSystem.Default.CreateFileInfo("..", FileAttributes.Directory, -1, DateTime.MinValue);
                else
                    item.FileInfo = _client.CreateFileInfo("..", FileAttributes.Directory, -1, DateTime.MinValue);
                viewList.Add(item);
            }

            if (local)
                _view.ListLocalFiles(viewList);
            else
                _view.ListRemoteFiles(viewList);
        }

        #endregion

        #region Local

        /// <summary>
        /// Refreshes the local list view.
        /// </summary>
        public void RefreshLocalList()
        {
            try
            {
                DirectoryInfo directory = new DirectoryInfo(_currentLocalDirectory);
                FileInfoCollection list = DiskFileSystem.Default.ListDirectory(directory.FullName);

                ListFiles(list, directory.FullName.Length <= 3, true);
            }
            catch (Exception exc)
            {
                _view.ShowError(exc);
                return;
            }
        }

        #endregion

        /// <summary>
        /// Changes local directory.
        /// </summary>
        /// <param name="dir">The new target path.</param>
        public void DoRemoteChangeDirectory(string dir)
        {
            if (_state != ConnectionState.Ready || dir == ".")
                return;

            if (dir == "..")
                dir = _client.GetDirectoryName(_currentDirectory);
            else
                dir = _client.CombinePath(_currentDirectory, dir);

            try
            {
                _client.SetCurrentDirectory(dir);
            }
            catch (Exception ex)
            {
                _view.ShowError(ex);
                return;
            }

            _view.UpdateRemotePath(dir);
            _currentDirectory = dir;
            RefreshRemoteList();
        }

        /// <summary>
        /// Changes local directory.
        /// </summary>
        /// <param name="dir">The new target path.</param>
        public void DoLocalChangeDirectory(string dir)
        {
            dir = Path.GetFullPath(Path.Combine(_currentLocalDirectory, dir));
            _view.UpdateLocalPath(dir);
            _currentLocalDirectory = dir;
            RefreshLocalList();
        }
    }
}
